File: plugin/src/test/java/com/stratio/cassandra/lucene/search/condition/DateRangeConditionTest.java
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.search.condition;

import com.stratio.cassandra.lucene.IndexException;
import com.stratio.cassandra.lucene.schema.Schema;
import com.stratio.cassandra.lucene.search.condition.builder.DateRangeConditionBuilder;
import org.apache.lucene.search.Query;
import org.apache.lucene.spatial.prefix.IntersectsPrefixTreeQuery;
import org.junit.Test;

import static com.stratio.cassandra.lucene.schema.SchemaBuilders.*;
import static com.stratio.cassandra.lucene.search.SearchBuilders.dateRange;
import static com.stratio.cassandra.lucene.search.condition.DateRangeCondition.*;
import static org.apache.lucene.spatial.query.SpatialOperation.*;
import static org.junit.Assert.*;

/**
 * @author Andres de la Pena {@literal <adelapena@stratio.com>}
 */
public class DateRangeConditionTest extends AbstractConditionTest {

    @Test
    public void testBuildString() {
        DateRangeConditionBuilder builder = new DateRangeConditionBuilder("field").boost(0.4)
                                                                                  .from("2015/01/05")
                                                                                  .to("2015/01/08")
                                                                                  .operation("intersects");
        DateRangeCondition condition = builder.build();
        assertNotNull("Condition is not built", condition);
        assertEquals("Boost is not set", 0.4f, condition.boost, 0);
        assertEquals("Field is not set", "field", condition.field);
        assertEquals("From is not set", "2015/01/05", condition.from);
        assertEquals("To is not set", "2015/01/08", condition.to);
        assertEquals("Operation is not set", "intersects", condition.operation);
    }

    @Test
    public void testBuildNumber() {
        DateRangeConditionBuilder builder = new DateRangeConditionBuilder("field").boost(0.4)
                                                                                  .from(1)
                                                                                  .to(2)
                                                                                  .operation("is_within");
        DateRangeCondition condition = builder.build();
        assertNotNull("Condition is not built", condition);
        assertEquals("Boost is not set", 0.4f, condition.boost, 0);
        assertEquals("Field is not set", "field", condition.field);
        assertEquals("From is not set", 1, condition.from);
        assertEquals("To is not set", 2, condition.to);
        assertEquals("Operation is not set", "is_within", condition.operation);
    }

    @Test
    public void testBuildDefaults() {
        DateRangeConditionBuilder builder = new DateRangeConditionBuilder("field");
        DateRangeCondition condition = builder.build();
        assertNotNull("Condition is not built", condition);
        assertNull("Boost is not set to default", condition.boost);
        assertEquals("Field is not set", "field", condition.field);
        assertEquals("From is not set to default", DEFAULT_FROM, condition.from);
        assertEquals("To is not set to default", DEFAULT_TO, condition.to);
        assertEquals("Operation is not set to default", DEFAULT_OPERATION, condition.operation);
    }

    @Test
    public void testJsonSerializationString() {
        DateRangeConditionBuilder builder = new DateRangeConditionBuilder("field").boost(0.4)
                                                                                  .from("from")
                                                                                  .to("to")
                                                                                  .operation("intersects");
        testJsonSerialization(builder,
                              "{type:\"date_range\",field:\"field\",boost:0.4," +
                              "from:\"from\",to:\"to\",operation:\"intersects\"}");
    }

    @Test
    public void testJsonSerializationNumber() {
        DateRangeConditionBuilder builder = new DateRangeConditionBuilder("field").boost(0.4)
                                                                                  .from("2015/01/05")
                                                                                  .to("2015/01/08")
                                                                                  .operation("contains");
        testJsonSerialization(builder,
                              "{type:\"date_range\",field:\"field\",boost:0.4," +
                              "from:\"2015/01/05\",to:\"2015/01/08\",operation:\"contains\"}");
    }

    @Test
    public void testJsonSerializationDefaults() {
        DateRangeConditionBuilder builder = new DateRangeConditionBuilder("field");
        testJsonSerialization(builder, "{type:\"date_range\",field:\"field\"}");
    }

    @Test
    public void testParseSpatialOperationIntersectsLowerCase() {
        assertEquals("Operation is not properly parsed", Intersects, parseSpatialOperation("intersects"));
    }

    @Test
    public void testParseSpatialOperationIntersectsUpperCase() {
        assertEquals("Operation is not properly parsed", Intersects, parseSpatialOperation("INTERSECTS"));
    }

    @Test
    public void testParseSpatialOperationIsWithinLowerCase() {
        assertEquals("Operation is not properly parsed", IsWithin, parseSpatialOperation("is_within"));
    }

    @Test
    public void testParseSpatialOperationIsWithinUpperCase() {
        assertEquals("Operation is not properly parsed", IsWithin, parseSpatialOperation("IS_WITHIN"));
    }

    @Test
    public void testParseSpatialOperationIContainsLowerCase() {
        assertEquals("Operation is not properly parsed", Contains, parseSpatialOperation("contains"));
    }

    @Test
    public void testParseSpatialOperationContainsUpperCase() {
        assertEquals("Operation is not properly parsed", Contains, parseSpatialOperation("CONTAINS"));
    }

    @Test(expected = IndexException.class)
    public void testParseSpatialOperationNull() {
        parseSpatialOperation(null);
    }

    @Test(expected = IndexException.class)
    public void testParseSpatialOperationEmpty() {
        parseSpatialOperation("");
    }

    @Test(expected = IndexException.class)
    public void testParseSpatialOperationBlank() {
        parseSpatialOperation(" ");
    }

    @Test
    public void testQuery() {
        Schema schema = schema().mapper("name", dateRangeMapper("from", "to").pattern("yyyyMMdd Z")).build();
        DateRangeCondition condition = dateRange("name").from("20160305 PST").to("20160405 PST").build();
        Query query = condition.query(schema);
        assertNotNull("Query is not built", query);
        assertEquals("Query type is wrong", IntersectsPrefixTreeQuery.class, query.getClass());
        assertEquals("Query is wrong",
                     "IntersectsPrefixTreeQuery(fieldName=name,queryShape=" +
                     "[2016-03-05T08 TO 2016-04-05T08:00:00.000],detailLevel=9,prefixGridScanLevel=7)",
                     query.toString());
    }

    @Test
    public void testQueryOpenStart() {
        Schema schema = schema().mapper("name", dateRangeMapper("from", "to").pattern("yyyyMMdd Z")).build();
        DateRangeCondition condition = dateRange("name").to("20160305 PST").build();
        Query query = condition.query(schema);
        assertNotNull("Query is not built", query);
        assertEquals("Query type is wrong", IntersectsPrefixTreeQuery.class, query.getClass());
        assertEquals("Query is wrong",
                     "IntersectsPrefixTreeQuery(fieldName=name,queryShape=" +
                     "[-292269054-12-02T16:47:04.192 TO 2016-03-05T08:00:00.000],detailLevel=9,prefixGridScanLevel=7)",
                     query.toString());
    }

    @Test
    public void testQueryOpenStop() {
        Schema schema = schema().mapper("name", dateRangeMapper("from", "to").pattern("yyyyMMdd Z")).build();
        DateRangeCondition condition = dateRange("name").from("20160305 PST").build();
        Query query = condition.query(schema);
        assertNotNull("Query is not built", query);
        assertEquals("Query type is wrong", IntersectsPrefixTreeQuery.class, query.getClass());
        assertEquals("Query is wrong",
                     "IntersectsPrefixTreeQuery(fieldName=name,queryShape=" +
                     "[2016-03-05T08 TO 292278994-08-17T07:12:55.807],detailLevel=9,prefixGridScanLevel=7)",
                     query.toString());
    }

    @Test(expected = IndexException.class)
    public void testQueryWithoutValidMapper() {
        Schema schema = schema().mapper("name", uuidMapper()).build();
        DateRangeCondition condition = dateRange("name").from(1L).to(2L).build();
        condition.query(schema);
    }

    @Test
    public void testToString() {
        DateRangeCondition condition = dateRange("name").from(1).to(2).operation("contains").boost(0.3).build();
        assertEquals("Method #toString is wrong",
                     "DateRangeCondition{boost=0.3, field=name, from=1, to=2, operation=contains}",
                     condition.toString());
    }
}

File: plugin/src/test/scala/com/stratio/cassandra/lucene/search/condition/DateRangeConditionTest.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.search.condition

import com.stratio.cassandra.lucene.common.GeoOperation
import com.stratio.cassandra.lucene.schema.SchemaBuilders._
import com.stratio.cassandra.lucene.search.SearchBuilders.dateRange
import com.stratio.cassandra.lucene.search.condition.DateRangeCondition._
import com.stratio.cassandra.lucene.search.condition.builder.DateRangeConditionBuilder
import org.apache.lucene.spatial.prefix.IntersectsPrefixTreeQuery
import org.junit.Assert._
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner

/**
  * @author Andres de la Pena `adelapena@stratio.com`
  */
@RunWith(classOf[JUnitRunner])
class DateRangeConditionTest extends AbstractConditionTest {

  test("BuildString") {
    val builder = new DateRangeConditionBuilder("field").boost(0.4f).from("2015/01/05")
      .to("2015/01/08")
      .operation(GeoOperation.INTERSECTS)
    val condition = builder.build
    assertNotNull("Condition is not built", condition)
    assertEquals("Boost is not set", 0.4f, condition.boost, 0)
    assertEquals("Field is not set", "field", condition.field)
    assertEquals("From is not set", "2015/01/05", condition.from)
    assertEquals("To is not set", "2015/01/08", condition.to)
    assertEquals("Operation is not set", "intersects", condition.operation)
  }

  test("BuildNumber") {
    val builder = new DateRangeConditionBuilder("field").boost(0.4f)
      .from(1)
      .to(2)
      .operation(GeoOperation.IS_WITHIN)
    val condition = builder.build
    assertNotNull("Condition is not built", condition)
    assertEquals("Boost is not set", 0.4f, condition.boost, 0)
    assertEquals("Field is not set", "field", condition.field)
    assertEquals("From is not set", 1, condition.from)
    assertEquals("To is not set", 2, condition.to)
    assertEquals("Operation is not set", "is_within", condition.operation)
  }

  test("BuildDefaults") {
    val builder = new DateRangeConditionBuilder("field")
    val condition = builder.build
    assertNotNull("Condition is not built", condition)
    assertNull("Boost is not set to default", condition.boost)
    assertEquals("Field is not set", "field", condition.field)
    assertEquals("From is not set to default", DEFAULT_FROM, condition.from)
    assertEquals("To is not set to default", DEFAULT_TO, condition.to)
    assertEquals("Operation is not set to default", DEFAULT_OPERATION, condition.operation)
  }

  test("JsonSerializationString") {
    val builder = new DateRangeConditionBuilder("field").boost(0.4f)
      .from("from")
      .to("to")
      .operation(GeoOperation.INTERSECTS)
    testJsonSerialization(builder,
      "{type:\"date_range\",field:\"field\",boost:0.4," +
        "from:\"from\",to:\"to\",operation:\"intersects\"}")
  }

  test("JsonSerializationNumber") {
    val builder = new DateRangeConditionBuilder("field").boost(0.4f)
      .from("2015/01/05")
      .to("2015/01/08")
      .operation(GeoOperation.CONTAINS)
    testJsonSerialization(builder,
      "{type:\"date_range\",field:\"field\",boost:0.4," +
        "from:\"2015/01/05\",to:\"2015/01/08\",operation:\"contains\"}")
  }

  test("JsonSerializationDefaults") {
    val builder = new DateRangeConditionBuilder("field")
    testJsonSerialization(builder, "{type:\"date_range\",field:\"field\"}")
  }

  test("Query") {
    val schemaVal = schema().mapper("name",
      dateRangeMapper("from", "to").pattern("yyyyMMdd Z")).build
    val condition = dateRange("name").from("20160305 PST").to("20160405 PST").build
    val query = condition.query(schemaVal)
    assertNotNull("Query is not built", query)
    assertEquals("Query type is wrong", classOf[IntersectsPrefixTreeQuery], query.getClass)
    assertEquals("Query is wrong",
      "IntersectsPrefixTreeQuery(fieldName=name,queryShape=" +
        "[2016-03-05T08 TO 2016-04-05T08:00:00.000],detailLevel=9,prefixGridScanLevel=7)",
      query.toString())
  }

  test("QueryOpenStart") {
    val schemaVal = schema().mapper("name",
      dateRangeMapper("from", "to").pattern("yyyyMMdd Z")).build
    val condition = dateRange("name").to("20160305 PST").build
    val query = condition.query(schemaVal)
    assertNotNull("Query is not built", query)
    assertEquals("Query type is wrong", classOf[IntersectsPrefixTreeQuery], query.getClass)
    assertEquals("Query is wrong",
      "IntersectsPrefixTreeQuery(fieldName=name,queryShape=" +
        "[-292269054-12-02T16:47:04.192 TO 2016-03-05T08:00:00.000],detailLevel=9,prefixGridScanLevel=7)",
      query.toString())
  }

  test("QueryOpenStop") {
    val schemaVal = schema().mapper("name",
      dateRangeMapper("from", "to").pattern("yyyyMMdd Z")).build
    val condition = dateRange("name").from("20160305 PST").build
    val query = condition.query(schemaVal)
    assertNotNull("Query is not built", query)
    assertEquals("Query type is wrong", classOf[IntersectsPrefixTreeQuery], query.getClass)
    assertEquals("Query is wrong",
      "IntersectsPrefixTreeQuery(fieldName=name,queryShape=" +
        "[2016-03-05T08 TO 292278994-08-17T07:12:55.807],detailLevel=9,prefixGridScanLevel=7)",
      query.toString())
  }

  test("QueryWithoutValidMapper") {
    val schemaVal = schema().mapper("name", uuidMapper()).build
    val condition = dateRange("name").from(1L).to(2L).build
    condition.query(schemaVal)
  }

  test("ToString") {
    val condition = dateRange("name").from(1).to(2).operation(GeoOperation.CONTAINS).boost(0.3f).build
    assertEquals("Method #toString is wrong",
      "DateRangeCondition{boost=0.3, field=name, from=1, to=2, operation=contains}",
      condition.toString)
  }
}
